### PR TITLE
gh-106307: Fix PyMapping_GetOptionalItemString()

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
The resulting pointer was not set to NULL if the creation of a temporary string object was failed.

The tests were also missed due to oversight.

### CODE REVIEW

File: Modules/_testcapi/util.h [hunk 1]
```diff
@@ -23,3 +23,5 @@
         assert(!PyErr_Occurred());          \
         return PyLong_FromSsize_t(_ret);    \
     } while (0)
+
+#define UNINITIALIZED_PTR ((void *)"uninitialized")
```
Comment (vstinner) [inline Modules/_testcapi/util.h]: In my previous tests for PyDict API, that I removed, for arguments set by ``&arg``, I initalized them with ``= Py_True``. But yeah, any value different than NULL is fine.

Using this macro is not obvious, can you please add a comment to explain why it's needed? Just explain how it should be used.

**Review by vstinner — APPROVED** [review]
LGTM. I just propose adding a comment on the new macro.

Does Valgrind or similar tool detect the usage of uninitialized memory?

File: Modules/_testcapi/util.h (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/_testcapi/util.h]: I extracted this part of the PR into separate PR #109014, because I am going to backport it.

**Review by vstinner — APPROVED** [review]
LGTM.

The _testcapi now intializes the variable used in the test:

```c
static PyObject *
mapping_getoptionalitemstring(PyObject *self, PyObject *args)
{
    PyObject *obj, *value = UNINITIALIZED_PTR;
```

**Comment by vstinner** [comment]
@erlend-aasland: Idea for the devguide: We should suggest to **always** initialized the ``&value`` parameter to NULL in case of errors. I just avoid any risk of using an uninitialized variable by misusing the C API. What do you think?

**Comment by erlend-aasland** [comment]
> @erlend-aasland: Idea for the devguide: We should suggest to **always** initialized the `&value` parameter to NULL in case of errors. I just avoid any risk of using an uninitialized variable by misusing the C API. What do you think?

I have a PR for this, and you already approved it twice 😃 https://github.com/python/devguide/pull/1128

**Comment by serhiy-storchaka** [comment]
There are exceptions, like `_PyEval_SliceIndex()`, where the output parameter should not be set in some cases.

**Comment by erlend-aasland** [comment]
Exceptions are fine; guidelines are for the general case, not the special case.

**Comment by vstinner** [comment]
> I have a PR for this, and you already approved it twice 😃 https://github.com/python/devguide/pull/1128

Alright, I should approve it 3 or 4 times 👍👍👍👍

**Pull request merged**